### DISCUSSION TITLE
API for new 6.2 source selection

### DISCUSSION THREAD

**Discussion by mesa-git** [danielhrisca/asammdf#493] [Q&A] [answered]
Hello,

Thanks for the great work with this library.

I've recently updated to the new asammdf 6.2 and i've noticed that the api call has changed.
You have introduced source selection for multiple signals with same name.
This is a great addition. But i find myself unable to clearly see how to use the new feature.

This is what i've been doing before:

```
existing_signals = [name for name in self.SigMapMdfNames.values() if name in self.mdf]
self.non_existing_signals = [name for name in self.SigMapMdfNames.values() if name not in self.mdf]
# Use the list of signals to retrieve the data from the MDF file
mdf_selected_signal_data = self.mdf.select(existing_signals)
```
 
But it gives the new error "_raise_on_mutiple_occurences" because i did not provide the source.
What i want to do is search for the number of sources before, and load the same number of signals from each source.

Something like:

```
number_of_sources = mdf.getSources()
for source is number_of_sources:
    source_name = mdf.getSourceName(source)
    existing_signals = [name for name in self.SigMapMdfNames.values() if name in self.mdf(source)]
    self.non_existing_signals = [name for name in self.SigMapMdfNames.values() if name not in self.mdf(source)]
    # Use the list of signals to retrieve the data from the MDF file
    mdf_selected_signal_data[source_name] = self.mdf.select(existing_signals)
    etc...
```

I've looked a bit around at different examples, but could not find something to work.
I've tried the whereis(occurences) example from [here](https://github.com/danielhrisca/asammdf/discussions/457), but i could not get it to work.

```
​occurrences​ ​=​ ​mdf​.​whereis​(​"my_signal"​)
​signals​ ​=​ ​mdf​.​select​(occurrences​)
```

It gives me 2 tuples something like ((642, 35849), (707, 35849)) which i assume the first value is the source, and the second value is the signal itself. But how do I get the name of the source as string (ex. "ECU1") in order to build my data with it.

Any suggestion would be greatly appreciated :)

**Comment by JulienGrv** [comment]
Hi @gitMesa 

I highly recommend that you re-read the documentation but I think this example can help you understand the general principle:

```python
from pathlib import Path

from asammdf import MDF

path = Path("my_mdf_file.mf4")
channels = ("name1", "name2", "name3")
source_name = "ECU1"

with MDF(path, channels=channels) as mdf:
    occurrences = tuple(
        (None, gp_idx, cn_idx)
        for name in channels
        for gp_idx, cn_idx in mdf.whereis(name, source_name=source_name)
    )
    signals = mdf.select(occurrences)

```

**Comment by JulienGrv** [comment, marked as answer]
Hi @gitMesa 

Following our discussion today I made these changes for you: #494 

Please mark this question as answered.

**Reply by mesa-git** [reply]
Thanks for the support.
